### PR TITLE
Implement Manifest V3 objects and deployment

### PR DESCRIPTION
**Pull request by schulzh**
Implements:
* objects and logic required for reading and writing CC API V3 manifests
* client interface and reactor implementation to apply the manifest
* resource matching for v3 packages API
* `pushManifestV3` operation, similar to the existing `pushManifest` operation

I hope that the `pushManifestV3` operation is complete (especially the "app readiness" checks), since I did not find a documentation of this process and had to reverse engineer it from the CLI.

I chose to pass a raw byte array as manifest in the client instead of an object, because that would've required to add the snakeyaml dependency to the client as well. Since the manifest is just passed to and applied by the CC anyway, I saw no benefit of passing as an object.

Unfortunately, I could not get a local CF instance set up for the integration tests. Due to licensing issues at my company I cannot use Virtualbox and setting it up on a local OpenStack "DevStack" instance wasn't working either. If requested, I can add integration tests, but I cannot ensure that they are working as intended.

Solves #999 and #1142 but not #1103 yet

### CODE REVIEW

**Comment by pivotal-david-osullivan** [comment]
Thanks for the PR @schulzh! Are you able to follow the link in the EasyCLA comment above to get the commit authorised 
✅  and satisfy the bot? 

SInce this is a significant addition, if you can add the integration tests I am happy to run them in a lab!

**Comment by schulzh** [comment]
Hi @pivotal-david-osullivan,

I've added integration tests. Could you please run them in your lab?
I've also completed the CLA. From my side this PR should be ready to merge, so I'd be happy to have it reviewed.

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/ApplicationsV3.java [hunk 1]
```diff
@@ -207,6 +207,14 @@ public interface ApplicationsV3 {
      */
     Mono<StartApplicationResponse> start(StartApplicationRequest request);
 
+    /**
+     * Makes the Restart Application request
+     *
+     * @param request the Restart Application request
+     * @return the response from the Restart Application request
+     */
+    Mono<RestartApplicationResponse> restart(RestartApplicationRequest request);
```
Comment (mheath) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/ApplicationsV3.java:216]: Even though it makes sense to put "restart" next to "start" and "stop", the convention in CFJC is to keep method names in alphabetical order.

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/resourcematch/_MatchedResource.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    @JsonProperty("size_in_bytes")
+    @Nullable
+    abstract Integer getSize();
+
+    /**
+     * The path relative to app root
+     */
+    @JsonProperty("path")
+    @Nullable
+    abstract String getPath();
```
Comment (mheath) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/resourcematch/_MatchedResource.java:59]: Put methods in alphabetical order.

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/SpacesV3.java [hunk 3]
```diff
@@ -87,4 +87,6 @@ public interface SpacesV3 {
      */
     Mono<UpdateSpaceResponse> update(UpdateSpaceRequest request);
 
+    Mono<ApplyManifestResponse> applyManifest(ApplyManifestRequest request);
```
Comment (mheath) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/SpacesV3.java:90]: Put method in alphabetical order.

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/_ApplyManifestRequest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * The response payload for the apply manifest operation
+ */
+@JsonSerialize
+@Value.Immutable
+abstract class _ApplyManifestRequest {
+    /**
+     * The manifest to apply
+     */
+    @JsonIgnore
+    abstract byte[] manifest();
```
Comment (mheath) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/_ApplyManifestRequest.java:33]: Rename method to `getManifest()`.

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/_ApplyManifestResponse.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * The response payload for the apply manifest operation
+ */
+@JsonDeserialize
+@Value.Immutable
+abstract class _ApplyManifestResponse {
+
+    /**
+     * The Job ID
+     */
+    public abstract Optional<String> jobId();
```
Comment (mheath) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/_ApplyManifestResponse.java:34]: Rename to `getJobId`.

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/spaces/ReactorSpacesV3.java [hunk 6]
```diff
@@ -103,4 +107,12 @@ public Mono<UpdateSpaceResponse> update(UpdateSpaceRequest request) {
             .checkpoint();
     }
 
+    @Override
+    public Mono<ApplyManifestResponse> applyManifest(ApplyManifestRequest request) {
```
Comment (mheath) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/spaces/ReactorSpacesV3.java:111]: Keep methods in alphabetical order.

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/AbstractClientV3Operations.java [hunk 7]
```diff
@@ -149,6 +150,21 @@ protected final <T> Mono<HttpClientResponseWithParsedBody<T>> postWithResponse(O
                 .response()
                 .parseBodyWithResponse(responseType));
     }
+    protected <T> Mono<HttpClientResponseWithParsedBody<T>> postRawWithResponse(byte[] requestPayload, String contentType, Class<T> responseType, Function<UriComponentsBuilder, UriComponentsBuilder> uriTransformer) {
```
Comment (mheath) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/AbstractClientV3Operations.java:153]: Methods in alphabetical order.

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Application.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An application manifest that captures some of the details of how an application is deployed.  See <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html">the manifest
+ * definition</a> for more details.
+ */
+@Value.Immutable
+abstract class _ManifestV3Application extends _ApplicationManifestCommon {
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Application.java:32]: Alphabetize methods names.

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Process.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+package org.cloudfoundry.operations.applications;
+
+
+import org.cloudfoundry.Nullable;
+import org.cloudfoundry.client.v3.processes.HealthCheckType;
+import org.immutables.value.Value;
+
+@Value.Immutable
+abstract class _ManifestV3Process {
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Process.java:25]: Alphabetize methods names.

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Service.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package org.cloudfoundry.operations.applications;
+
+
+import org.cloudfoundry.Nullable;
+import org.immutables.value.Value;
+
+@Value.Immutable
+abstract class _ManifestV3Service {
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Service.java:24]: Alphabetize methods names.

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Sidecar.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+package org.cloudfoundry.operations.applications;
+
+
+import org.cloudfoundry.Nullable;
+import org.immutables.value.Value;
+
+import java.util.List;
+
+@Value.Immutable
+abstract class _ManifestV3Sidecar {
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_ManifestV3Sidecar.java:26]: Alphabetize methods.

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_Route.java [hunk 12]
```diff
@@ -33,4 +34,11 @@
     @JsonProperty("route")
     abstract String getRoute();
 
+    /**
+     * The protocol
+     */
+    @JsonProperty("protocol")
+    @Nullable
+    abstract ManifestV3RouteProtocol getProtocol();
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/_Route.java:42]: Keep methods in alphabetical order.

File: cloudfoundry-util/src/main/java/org/cloudfoundry/util/ResourceMatchingUtilsV3.java [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Enumeration;
+import java.util.List;
+
+/**
+ * Utilities for matching resources with objects for v3 API
+ */
+public final class ResourceMatchingUtilsV3 {
```
Comment (mheath) [inline cloudfoundry-util/src/main/java/org/cloudfoundry/util/ResourceMatchingUtilsV3.java:46]: Methods in alphabetical order

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/Applications.java [hunk 14]
```diff
@@ -138,6 +138,8 @@ public interface Applications {
      */
     Mono<Void> pushManifest(PushApplicationManifestRequest request);
 
+    Mono<Void> pushManifestV3(PushManifestV3Request request);
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/Applications.java:141]: Add JavaDoc

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/DefaultApplications.java [hunk 15]
```diff
@@ -431,6 +461,142 @@ public Mono<Void> pushManifest(PushApplicationManifestRequest request) {
...
+            .flatMap(function((cloudFoundryClient, spaceId, manifestApp) ->
+                getApplicationIdV3(cloudFoundryClient, manifestApp.getName(), spaceId)
+                    .flatMap(appId -> Mono.zip(Mono.just(appId), createPackage(cloudFoundryClient, appId, manifestApp)))
+                    .flatMap(function((appId, packageId) ->
+                        buildAndStage(cloudFoundryClient, manifestApp, packageId)
+                            .flatMap(dropletId -> applyDropletAndWaitForRunning(cloudFoundryClient, manifestApp.getName(), appId, dropletId))))))
+            .then();
+    }
+
+    private static Mono<Void> applyDropletAndWaitForRunning(CloudFoundryClient cloudFoundryClient, String appname, String appId, String dropletId) {
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/DefaultApplications.java:489]: Move private methods to bottom of class and make static (same style as the other privet methods)

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/DefaultApplications.java [hunk 16]
```diff
@@ -1426,6 +1592,14 @@ private static Flux<PrivateDomainResource> requestListPrivateDomains(CloudFoundr
                     .build()));
     }
 
+    private static Flux<DomainResource> requestListDomains(CloudFoundryClient cloudFoundryClient, String organizationId) {
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/DefaultApplications.java:1595]: Method in alphabetical order

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/DefaultApplications.java [hunk 17]
```diff
@@ -1985,6 +2159,31 @@ private static Mono<Void> waitForRunning(CloudFoundryClient cloudFoundryClient,
             .then();
     }
 
+    private static Mono<Void> waitForV3Running(CloudFoundryClient cloudFoundryClient, String applicationName, String applicationId, Duration startupTimeout) {
```
Comment (mheath) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/DefaultApplications.java:2162]: Rename to `waitForRunningV3`

File: integration-test/src/test/java/org/cloudfoundry/client/v3/ResourceMatchTest.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+public class ResourceMatchTest extends AbstractIntegrationTest {
+
+    @Autowired
+    private CloudFoundryClient cloudFoundryClient;
+
+    @Autowired
+    private Mono<String> spaceId;
+
+    //TODO how to check if resource matching is enabled on this CF instance?
+    @IfCloudFoundryVersion(greaterThanOrEqualTo = CloudFoundryVersion.UNSPECIFIED) //TODO how to select this version?
```
Comment (mheath) [inline integration-test/src/test/java/org/cloudfoundry/client/v3/ResourceMatchTest.java:59]: I will look into this

File: integration-test/src/test/java/org/cloudfoundry/client/v3/SpacesTest.java [hunk 19]
```diff
@@ -48,10 +50,13 @@
 import org.junit.Ignore;
 import org.junit.Test;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.ClassPathResource;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.test.StepVerifier;
+import sun.misc.IOUtils;
```
Comment (mheath) [inline integration-test/src/test/java/org/cloudfoundry/client/v3/SpacesTest.java:57]: Don't use the `sun.misc` package.

You could use `org.springframework.util.StreamUtils` and change line 279 to something like:
```
byte[] manifest = StreamUtils.copyToByteArray(new ClassPathResource("test-manifest.yml").getInputStream());
```

Or you could do something like:
```
 File file = new ClassPathResource("test-manifest.yml").getFile();
 byte[] manifest = Files.readAllBytes(file.toPath());
```

**Review by mheath — CHANGES_REQUESTED** [review]
This is an amazing contribution. Thank you thank you thank you for all your hard work.

There's a fix that needs to happen in v3/SpacesTest.java (see the comments there)

Other than that, most of the rest of my comments are nitpicks about method order.

Once you fix these things, I'll merge this into its own branch so that I can write acceptance tests for it. Once we have tests, we'll merge this into `main`.

**Comment by mheath** [comment]
Will you change the branch from `main` to `manifest-v3` on this PR and I will get it merged and then merge in my acceptance tests?

**Comment by schulzh** [comment]
Hi @mheath,

I've changed the base branch. It looks like its lagging behind main a few hounded commits. Should I rebase it?

**Comment by schulzh** [comment]
Hello @mheath,

do you have any updates on this? What is missing to get it merged to master?

**Pull request merged**